### PR TITLE
stark: wall-shear prediction in local surface tangent frame

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The wall-shear y/z gap (tau_y: 13.47 vs AB-UPT 3.65, tau_z: 14.52 vs 3.63 — both ~4× off) is caused in part by the model learning wall-shear stress in world coordinates (x, y, z), where the physical constraint (no-slip BC forces tau_n = 0) is implicit and the axes y/z have low-magnitude, high-frequency variation that is hard to regress directly.

**Hypothesis:** If we predict wall-shear stress in the local surface tangent frame {t1, t2, n} — where n is the outward surface normal and (t1, t2) are two orthogonal tangent vectors per point — then:
1. The model no longer needs to regress a near-zero tau_n component against noise
2. The physically meaningful in-plane components (tau_t1, tau_t2) carry the full signal
3. The loss landscape is better conditioned because the full magnitude is concentrated in two channels instead of being distributed across three world-frame axes

The surface normals are already available at `surface_x[..., 3:6]`. We rotate the prediction and target into the surface frame, compute the loss in frame coordinates, and rotate the prediction back to world coordinates for validation.

Reference: Surface-tangent frame prediction is standard in aerodynamic drag/separation prediction literature (Bonnet et al., 2022; Fortunato et al., 2022). The key property: in the local frame, tau_n is identically zero by the Newtonian no-slip BC, so training in frame coordinates implicitly enforces the physical constraint.

## Instructions

**Overview:** Add a `--use-tangent-frame-loss` flag that rotates the wall-shear prediction and target into the local {t1, t2, n} frame before computing loss. Training predicts in the tangent frame; validation converts back to world coords for consistent metric reporting.

**Step 1: Build the tangent frame rotation.**

Add a function `build_tangent_frame(normals)` near the existing `project_tangential()` function in `train.py`:

```python
def build_tangent_frame(normals: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
    """Build orthonormal surface frame {t1, t2, n} per point.
    
    normals: [B, N, 3] unit outward normals
    Returns: (t1, t2, n_hat) each [B, N, 3] in fp32.
    
    Uses Frisvad/Duff et al. (2017) method for a numerically stable
    orthonormal basis from a single normal vector, avoiding cross-product
    singularities near the poles.
    """
    n = F.normalize(normals.float(), dim=-1, eps=1e-8)
    # Duff et al. 2017 — sign-based Frisvad
    # Avoids the singularity at nz ≈ -1 that naive Gram-Schmidt hits
    nz = n[..., 2:3]
    sign = torch.where(nz >= 0, torch.ones_like(nz), -torch.ones_like(nz))
    a = -1.0 / (sign + nz)
    b = n[..., 0:1] * n[..., 1:2] * a
    t1 = torch.cat([
        1.0 + sign * n[..., 0:1].square() * a,
        sign * b,
        -sign * n[..., 0:1]
    ], dim=-1)
    t2 = torch.cat([
        b,
        sign + n[..., 1:2].square() * a,
        -n[..., 1:2]
    ], dim=-1)
    return t1, t2, n
```

**Step 2: Add rotation helpers.**

```python
def world_to_frame(vec: torch.Tensor, t1: torch.Tensor, t2: torch.Tensor, n_hat: torch.Tensor) -> torch.Tensor:
    """Rotate [B, N, 3] vec from world to local tangent frame (t1, t2, n). fp32."""
    v = vec.float()
    return torch.stack([
        (v * t1).sum(-1),
        (v * t2).sum(-1),
        (v * n_hat).sum(-1),
    ], dim=-1)

def frame_to_world(vec: torch.Tensor, t1: torch.Tensor, t2: torch.Tensor, n_hat: torch.Tensor) -> torch.Tensor:
    """Rotate [B, N, 3] vec from local tangent frame back to world coords. fp32."""
    v = vec.float()
    return v[..., 0:1] * t1 + v[..., 1:2] * t2 + v[..., 2:3] * n_hat
```

**Step 3: Add the config flag and modify `train_loss()`.**

In `TrainConfig`:
```python
use_tangent_frame_loss: bool = False  # predict tau in local {t1, t2, n} frame
```

In `train_loss()`, add `use_tangent_frame_loss: bool = False` to the signature and in the loss computation block, after the existing tangential-projection block:

```python
if use_tangent_frame_loss:
    normals = batch.surface_x[..., 3:6]
    ws_std = transform.surface_y_std[1:4]
    ws_mean = transform.surface_y_mean[1:4]
    # Denormalize to physical space
    ws_pred_phys = surface_pred_norm[..., 1:4] * ws_std + ws_mean
    ws_true_phys = surface_target[..., 1:4] * ws_std + ws_mean
    # Build per-point tangent frame
    t1, t2, n_hat = build_tangent_frame(normals)
    # Rotate into local frame
    ws_pred_frame = world_to_frame(ws_pred_phys, t1, t2, n_hat).to(ws_pred_phys.dtype)
    ws_true_frame = world_to_frame(ws_true_phys, t1, t2, n_hat).to(ws_true_phys.dtype)
    # The n-component should be near zero (no-slip BC); include it in loss to penalize violations
    # Renormalize in frame space using the SAME std/mean (isotropic rescaling)
    ws_pred_frame_norm = (ws_pred_frame - ws_mean) / ws_std
    ws_true_frame_norm = (ws_true_frame - ws_mean) / ws_std
    surface_pred_used = torch.cat([surface_pred_norm[..., :1], ws_pred_frame_norm], dim=-1)
    surface_target_used = torch.cat([surface_target[..., :1], ws_true_frame_norm], dim=-1)
    # Log the mean n-component as a diagnostic (should decrease toward 0 during training)
    if bool(batch.surface_mask.any()):
        n_comp_rms = float(ws_pred_frame[..., 2][batch.surface_mask].abs().mean().detach().cpu().item())
    else:
        n_comp_rms = float("nan")
```

Add `metrics["wallshear_pred_ncomp_rms"] = n_comp_rms` to log the n-component magnitude.

**Step 4: Add CLI flags.**

```python
parser.add_argument("--use-tangent-frame-loss", action="store_true", default=False,
    help="Compute wall-shear loss in local surface tangent frame {t1,t2,n}")
```

**Step 5: Validation.** The val loop must convert predictions BACK to world coordinates before computing the existing `wall_shear_y_rel_l2_pct` metrics (which are in world coords). At the val inference call, if `use_tangent_frame_loss`, after getting `surface_preds`:

```python
# If trained in tangent frame, rotate back to world for validation
if config.use_tangent_frame_loss:
    ws_pred_norm = surface_preds[..., 1:4]
    ws_pred_phys_frame = ws_pred_norm * ws_std + ws_mean  # unnorm from frame-space
    t1, t2, n_hat = build_tangent_frame(normals)
    ws_pred_phys_world = frame_to_world(ws_pred_phys_frame, t1, t2, n_hat).to(ws_pred_norm.dtype)
    ws_pred_world_norm = (ws_pred_phys_world - ws_mean) / ws_std
    surface_preds = torch.cat([surface_preds[..., :1], ws_pred_world_norm], dim=-1)
```

Note: This is necessary to ensure validation metrics are in world coordinates and comparable to the baseline.

**Step 6: Run a 2-arm sweep.**

Use `--wandb_group stark-tangent-frame` for all runs.

**Arm A (tangent frame, lr=5e-4):**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --lr 5e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 \
  --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 \
  --wallshear-z-weight 2.0 \
  --pos-max-wavelength 1000 \
  --lr-warmup-steps 500 \
  --use-tangent-frame-loss \
  --seed 42 \
  --wandb_group stark-tangent-frame \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

**Arm B (tangent frame + lr=3e-4, stability test):**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py \
  ... (same as above but) \
  --lr 3e-4 \
  --seed 42 \
  --wandb_group stark-tangent-frame
```

**Arm C (tangent frame, lr=5e-4, seed=1 confirmation if A survives):**
Launch on GPU 2 after confirming Arm A passes step 5000 without divergence.

**Arm D (baseline, no tangent frame — control arm):**
```bash
CUDA_VISIBLE_DEVICES=3 python train.py \
  ... (same as Arm A but without --use-tangent-frame-loss) \
  --seed 42 \
  --wandb_group stark-tangent-frame
```

Arm D is the matched control. If the only difference from baseline is `--use-tangent-frame-loss`, any gap between Arm A/B and Arm D is cleanly attributable to the tangent frame.

## Baseline

Current best — PR #183 (fern), W&B run `bplngfyo`:
```
val_primary/abupt_axis_mean_rel_l2_pct: 10.21   ← merge bar (must beat)
val_primary/surface_pressure_rel_l2_pct: 6.85
val_primary/wall_shear_rel_l2_pct: 11.43
val_primary/volume_pressure_rel_l2_pct: 6.32
val_primary/wall_shear_x_rel_l2_pct: 9.89
val_primary/wall_shear_y_rel_l2_pct: 13.47      ← primary target (~4× AB-UPT)
val_primary/wall_shear_z_rel_l2_pct: 14.52      ← primary target (~4× AB-UPT)
```

AB-UPT reference targets: wall_shear_y=3.65, wall_shear_z=3.63.

Reproduce baseline:
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 5e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --pos-max-wavelength 1000 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

## Key Diagnostics to Report

Per epoch, per arm:
1. `val_primary/wall_shear_y_rel_l2_pct` and `wall_shear_z_rel_l2_pct` — the primary hypothesis target
2. `val_primary/abupt_axis_mean_rel_l2_pct` — merge metric
3. `train/wallshear_pred_ncomp_rms` — should trend toward 0 (confirms no-slip BC is being learned)
4. `train/grad_pre` median+max — stability watchdog

**Stability note:** If `grad_pre` > 50 sustained or spikes > 500, kill the arm immediately and report. The tangent-frame rotation adds a fp32 compute path that is numerically stable, but the normalized-space round-trip (phys → frame → norm) could alter the gradient landscape. Arm B (lr=3e-4) is the fallback if Arm A diverges.

**Merge bar:** val_abupt < 10.21 from best-val checkpoint.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: stark — your Round 14 assignment (PR #227, wall-shear prediction in local surface tangent frame) was posted on 2026-05-01. Zero comments since then. Please confirm you have launched the experiment. This is the most physically-motivated approach we have for the tau_y/z gap: predicting shear in the local {t1, t2, n} frame directly enforces the no-slip BC geometry and should disentangle the y/z components from the global coordinate system that the model currently struggles with. Confirm launch or post deviation note.

**Comment by morganmcg1** [comment]
**ADVISOR:** Flagged to human team via Issue #248 — the advisor pod lacks RBAC permissions to create Kubernetes ConfigMaps or Deployments, so the `senpai-yi-stark` executor pod cannot be provisioned from within this pod.

A human operator needs to run the following from a machine with cluster admin access:

```bash
cd /workspace/senpai/k8s
python launch.py \
  --tag yi \
  --target_repo_url https://github.com/morganmcg1/DrivAerML \
  --names stark \
  --gpus_per_student 4 \
  --cpu_per_gpu 15 \
  --memory_gi_per_gpu 120 \
  --wandb_project senpai-v1-drivaerml \
  --advisor_branch yi \
  --max_epochs 999 \
  --timeout_minutes 360
```

Once the pod is running, pick up PR #227 and implement the surface tangent frame approach. Standing by.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #227 because ADVISOR (2026-05-02): Closing this PR per human operator directive on Issue #248.

Operator confirmed (2026-05-01 23:05Z) that the senpai-yi-stark pod is absent — no deployment, no ConfigMap, no .claude/ logs to recover. The stark slot in the yi fleet does not exist (current active yi fleet: alphonse, askeladd, chihiro, edward, emma, fern, frieren, gilbert, haku, kohaku, nezuko, norman, senku, tanjiro, thorfinn, violet — 16 students; no stark). The operator's recommendation: reassign rather than wait on provisioning.

The surface tangent-frame hypothesis (predict wall shear in local {t1, t2, n} frame to enforce no-slip BC geometry and disentangle tau_y/z from global coords) remains the most physically-motivated approach in our pending queue for the tau_y/z gap. It will be reassigned to the next available yi student via a fresh PR off the current best baseline (PR #222, val_abupt = 9.291%).

All 16 active yi students currently have status:wip PRs in flight, so no immediate idle slot exists. Will pick this up on the next round when a student becomes idle (likely after the violet/askeladd/fern/haku sweeps wrap)..
